[RFC] [stdlib] Unified code base of Darwin and Glibc

rintaro opened the pull request:
#### What's in this pull request?

In this PR, at least for now, I don't intend to implement @modocache's proposal: [Unify "import Darwin/Glibc" to simply "Libc"](http://thread.gmane.org/gmane.comp.lang.swift.evolution/8905/),
but just to make code management easier for future enhancement. No API changes.

Unified `stdlib/public/SDK/Darwin` and `stdlib/public/Glibc` into single
`stdlib/public/Platform` directory, and share the code where possible.
The naming of the directory name **`Platform`** is still debatable. (`System`, `Host`, `Libc`, ...)

Using this, I implemented `tgmath` in Glibc module. (included in this PR)
Actually, the original motivation is to implement tgmath.
We have several options to do that:
- Copy `Darwin/tgmath.swift.gyb` to `Glibc/tgmath.swift.gyb`.  https://github.com/apple/swift/compare/master...rintaro:linux-tgmath
- Symlink `tgmath.swift.gyb` https://github.com/tchagnon/swift/commit/742eff3914f2a62e799980523badaca4eeee95fe
- Unify Darwin and Glibc (this PR)

IMO, the last option is the best for future enhancement.

Please tell me if we need discussion on swift-dev ML.

#### Resolved bug number: ([SR-638](https://bugs.swift.org/browse/SR-638))
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Decide to unify them or not.
- [ ] Decide the naming of the directory.
- [ ] Test pull request on Swift continuous integration.

rintaro commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -0,0 +1,82 @@
+add_swift_library(${swift_platform_name} IS_SDK_OVERLAY
+    ${swift_platform_sources}
+    ${swift_platform_flags})
+
```
Removed `INSTALL_IN_COMPONENT stdlib-experimental` because it's [eventually reset to `sdk-overlay`](https://github.com/apple/swift/blob/da60535cedd58c7f6726cb14e821f4473fc214ae/cmake/modules/AddSwift.cmake#L1365).

gribozavr commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+    ${swift_platform_sources}
+    ${swift_platform_flags})
+
+set(glibc_modulemap_targat_list)
```
Unused variable?

rintaro replied:
Oops typo, will fix.

gribozavr commented:
Generally the approach looks good to me.  This is a great improvement for Linux and FreeBSD overlays!

Could you resolve the conflicts?  Please make sure to preserve changes from https://github.com/apple/swift/pull/1724.

rintaro commented:
@gribozavr Thanks for the review!
Rebased on 1e4f7109fa36117a679afd3df2847e0619bdbc89 (Merge pull request #1724 from compnerd/correct-linkage).
Fixed typo in `CMakeLists.txt`.
Tested with `build-script -RT` (Mac, Linux), and `build-script --preset buildbot_linux` (Linux)

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@rintaro Seems like the buildbot found some issues on Linux...

rintaro commented on stdlib/public/Platform/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+          #        This will break if cross-compiling from a distro that uses the
+          #        architecture as part of the path to a distro that does not.
+          set(GLIBC_ARCH_INCLUDE_PATH
+              "${GLIBC_INCLUDE_PATH}/${CMAKE_LIBRARY_ARCHTECTURE}")
```
Awful typo. `CMAKE_LIBRARY_ARCHTECTURE ` -> `CMAKE_LIBRARY_ARCHITECTURE `

gribozavr commented:
@swift-ci Please test

rintaro commented:
Thanks!

gribozavr commented:
@rintaro Thank you for removing all that duplication!
